sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/errorhandling/ErrorHandler.java

```diff
@@ -0,0 +1,208 @@
+ * Error Handlers must be closed before a pipeline is run to properly pipe error collections to the
+ * sink, and the pipeline will be rejected if any handlers aren't closed.
+ *
+ * @param <ErrorT> The type of the error object. This will usually be a {@link BadRecord}, but can
+ *     be any type
+ * @param <OutputT> The return type of the sink PTransform.
+ *     <p>Usage of Error Handlers:
+ *     <p>Simple usage with one DLQ
+ *     <pre>{@code
+ * PCollection<?> records = ...;
+ * try (ErrorHandler<E,T> errorHandler = pipeline.registerErrorHandler(SomeSink.write())) {
+ *  PCollection<?> results = records.apply(SomeIO.write().withDeadLetterQueue(errorHandler));
+ * }
+ * results.apply(SomeOtherTransform);
+ * }</pre>
+ *     Usage with multiple DLQ stages
+ *     <pre>{@code
+ * PCollection<?> records = ...;
+ * try (ErrorHandler<E,T> errorHandler = pipeline.registerErrorHandler(SomeSink.write())) {
+ *  PCollection<?> results = records.apply(SomeIO.write().withDeadLetterQueue(errorHandler))
+ *                        .apply(OtherTransform.builder().withDeadLetterQueue(errorHandler));
+ * }
+ * results.apply(SomeOtherTransform);
+ * }</pre>
+ */
+public interface ErrorHandler<ErrorT, OutputT extends POutput> extends AutoCloseable {
+
+  void addErrorCollection(PCollection<ErrorT> errorCollection);
+
+  boolean isClosed();
+
+  OutputT getOutput();
+
+  class PTransformErrorHandler<ErrorT, OutputT extends POutput>
+      implements ErrorHandler<ErrorT, OutputT> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(PTransformErrorHandler.class);
+    private final PTransform<PCollection<ErrorT>, OutputT> sinkTransform;
+
+    private final Pipeline pipeline;
+
+    private final List<PCollection<ErrorT>> errorCollections = new ArrayList<>();
+
+    private @Nullable OutputT sinkOutput = null;
+
+    private boolean closed = false;
+
+    /**
+     * Constructs a new ErrorHandler, but should not be called directly. Instead, call
+     * pipeline.registerErrorHandler to ensure safe pipeline construction
+     */
+    @Internal
+    public PTransformErrorHandler(
+        PTransform<PCollection<ErrorT>, OutputT> sinkTransform, Pipeline pipeline) {
+      this.sinkTransform = sinkTransform;
+      this.pipeline = pipeline;
+    }
+
+    @Override
+    public void addErrorCollection(PCollection<ErrorT> errorCollection) {
+      errorCollections.add(errorCollection);
+    }
+
+    @Override
+    public boolean isClosed() {
+      return closed;
+    }
+
+    @Override
+    public OutputT getOutput() {
+      if (!this.isClosed()) {
+        throw new IllegalStateException(
+            "ErrorHandler must be finalized before the output can be returned");
+      }
+      // make the static analysis checker happy
+      Preconditions.checkArgumentNotNull(sinkOutput);
+      return sinkOutput;
+    }
+
+    @Override
+    public void close() {
+      closed = true;
```

Comment:
We could require that a Coder<ErrorT> be provided. The BadRecordsHandler always has a known error types.

Reply:
Thats true, but it would only be needed in the relatively rare case where someone configures an error handler, but then doesn't use it. I think another option would be to simply raise an error, as this probably means something is misconfigured.

Reply:
I'm not sure we want to have the user provide a coder, to keep the interface simple.

We could change registration to Pipeline::registerBadRecordHandler, and just keep that logic in the pipeline. That has the side benefit of letting us know more about specific error handlers for the future case where we support other kinds of error handling